### PR TITLE
Form with truthy expression displayed despite permissions

### PR DESCRIPTION
**Pull request by njogz**
# Description

This PR reorders the checks done before returning a form to take care of the bug below.

medic/cht-core#7021

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/services/xml-forms.service.ts [hunk 1]
```diff
@@ -162,6 +162,9 @@ export class XmlFormsService {
       if (!validSoFar) {
         return false;
       }
+      if (form.context.permission) {
+        return this.authService.has(form.context.permission);
```
Comment (dianabarsan) [inline webapp/src/ts/services/xml-forms.service.ts:166]: Returning here means that if we have a permission, we never check the expression?

File: webapp/tests/karma/ts/services/xml-forms.service.spec.ts [hunk 2]
```diff
@@ -777,6 +777,43 @@ describe('XmlForms service', () => {
 
     });
 
+    it('does not return a form with a truthy expression if the user does not have relevant permissions', () => {
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/services/xml-forms.service.spec.ts:780]: Please add a unit test where a form has a permission and an expression and: 
1. user has permission and expression is valid
2. user doesn't have permission and expression is valid
3. user has permission and expression is invalid.

**Review by dianabarsan — CHANGES_REQUESTED** [review]

**Comment by njogz** [comment]
Thanks for the review @dianabarsan. I have addressed your comments please have another look.

File: webapp/src/ts/services/xml-forms.service.ts [hunk 3]
```diff
@@ -164,7 +164,16 @@ export class XmlFormsService {
       }
       if (form.context.expression) {
         try {
-          return this.evaluateExpression(form.context.expression, options.doc, user, options.contactSummary);
+          const evaluatedExpression = this.evaluateExpression(
```
Comment (dianabarsan) [inline webapp/src/ts/services/xml-forms.service.ts:167]: I'm wondering if we can change this so we only ever call `evaluateExpression` and `authService.has` once per form. 
For example, extracting them into separate functions like: 
```
private checkFormExpression(form, doc, user, contactSummary) {
    if (!form.context.expression) {
      return true;
    }

    try {
      return this.evaluateExpression(
        form.context.expression,
        doc,
        user,
        contactSummary
      );
    } catch(err) {
      console.error(`Unable to evaluate expression for form: ${form._id}`, err);
      return false;
    }
  }

  private checkFormPermissions(form) {
    if (!form.context.permission) {
      return true;
    }

    return this.authService.has(form.context.permission);
  }
```
And then doing:  
```
return this
      .filterContactTypes(form.context, options.doc)
      .then(valid => valid && this.checkFormPermissions(form))
      .then(valid => valid && this.checkFormExpression(form, options.doc, user, options.contactSummary));
```

How does that sound?

**Review by dianabarsan — COMMENTED** [review]
Thanks a lot for adding the tests! I left one comment inline about reduced duplication.

File: webapp/src/ts/services/xml-forms.service.ts (hunk 3 above)
Comment (njogz) [reply webapp/src/ts/services/xml-forms.service.ts]: Brilliant! I have made the refactor as suggested.

**Review by dianabarsan — APPROVED** [review]

**Comment by dianabarsan** [comment]
Great work!

**Comment by jkuester** [comment]
@njogz after consulting with @craig-landry we have decided to go ahead and merge this today (since it has [passed AT](https://github.com/medic/cht-core/issues/7021#issuecomment-1021744715)) so that we can begin the `3.14.0` release process.  If this is done in error, please let me know and we can fix things!

**Pull request merged**